src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,310 @@
+            AD_CONTRACT_CHECK(sparqlExpression::isConstantResult<T>);
+            if constexpr (isStrongId) {
+              return singleResult;
+            } else if constexpr (sparqlExpression::isConstantResult<T>) {
+              return sparqlExpression::detail::constantExpressionResultToId(
+                  AD_FWD(singleResult), localVocab_);
+            } else {
+              // This should never happen since aggregates always return
+              // constants.
+              AD_FAIL();
+            }
+          },
+          std::move(expressionResult));
+    }
+  }
+  resetAggregationData();
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+void LazyGroupBy<NUM_AGGREGATED_COLS>::processNextBlock(
+    sparqlExpression::EvaluationContext& evaluationContext, size_t beginIndex,
+    size_t endIndex) {
+  size_t blockSize = endIndex - beginIndex;
+  evaluationContext._beginIndex = beginIndex;
+  evaluationContext._endIndex = endIndex;
+
+  for (auto& aggregateAlias : aggregateAliases_) {
+    for (auto& aggregate : aggregateAlias.aggregateInfo_) {
+      // Evaluate child expression on block
+      auto exprChildren = aggregate.expr_->children();
+      sparqlExpression::ExpressionResult expressionResult =
+          exprChildren[0]->evaluate(&evaluationContext);
+
+      std::visit(
+          [this, blockSize, &evaluationContext,
+           &aggregate]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) {
+            auto generator = sparqlExpression::detail::makeGenerator(
+                std::forward<T>(singleResult), blockSize, &evaluationContext);
+
+            for (const auto& val : generator) {
+              addToAggregateFunction(aggregate.aggregateDataIndex_, val,
+                                     &evaluationContext);
+            }
+          },
+          std::move(expressionResult));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+
+template <size_t NUM_AGGREGATED_COLS>
+void LazyGroupBy<NUM_AGGREGATED_COLS>::populateGroupBlock(
+    const IdTable& idTable, size_t row) {
+  if (currentGroupBlock_.empty()) {
+    for (size_t col : groupByCols_) {
+      currentGroupBlock_.emplace_back(col, idTable(0, col));
+    }
+  } else {
+    for (auto& columnPair : currentGroupBlock_) {
+      columnPair.second = idTable(row, columnPair.first);
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+bool LazyGroupBy<NUM_AGGREGATED_COLS>::rowMatchesCurrentBlock(
+    const IdTable& idTable, size_t row) {
+  return std::all_of(currentGroupBlock_.begin(), currentGroupBlock_.end(),
+                     [&](const auto& columns) {
+                       return idTable(row, columns.first) == columns.second;
+                     });
+}
```

Comment:
I think we should use a column-based
approach, that directly computes ALL block borders
(see one of my new `Distinct` functions in `Engine.h/.cpp`